### DISCUSSION TITLE
Help with providing good error messages to the user

### DISCUSSION THREAD

**Discussion by Elias-Graf** [lalrpop/lalrpop#1012] [Q&A]
I'm trying to write a parser for the following code:

```yuck
(metric 
    :foo "bar"
    :label "123"
    :value 1
    "hello, world!"
    (bar))
```

For that, I have the following parsers:

```
pub WidgetCall: Spanned<WidgetCall> = {
    <l:@L> "(" <name:Symbol> <args:WidgetCallArg*> <children:WidgetCallChild*> ")" <r:@R> => {
        (l, WidgetCall::new(name, args, children), r).into()
    },
};

pub WidgetCallArg: WidgetCallArg = {
    ":" <name:Symbol> <value:Atom> => WidgetCallArg::new(name, value),
};

pub WidgetCallChild: WidgetCallChild = {
    Atom => <>.into(),
    WidgetCall => <>.into(),
};
```

The main goal, I'm pursuing for the parser, is to provide rich error message, currently for the `WidgetCallArg`. I would like that almost any way an arg is provided, there should be a message, telling you, what you are missing. E.g. missing the colon, missing the argument name, or the value.

Adding a rule like this:

```
pub WidgetCallArg: WidgetCallArg = {
    ":" <name:Symbol> <value:Atom> => WidgetCallArg::new(name, value),
    ":" <name:Symbol> => todo!("missing arg"),
};
```

It is problematic, because the program now can't decide if the next atom should be the argument, or the `WidgetCallChild` (which can also just be an atom).

So I've tried to work around this, by saying "you can't have children if the value of the argument is missing". Translating to the following code:

```
pub WidgetCall: Spanned<WidgetCall> = {
    <l:@L> "(" <name:Symbol> <args:WidgetCallArg*> <children:WidgetCallChild*> ")" <r:@R> => {
        (l, WidgetCall::new(name, args, children), r).into()
    },
    <l:@L>
    "("
    <name:Symbol>
    <mut args:WidgetCallArg*>
    <arg_trail:WidgetCallArgMissingValue>
    ")"
    <r:@R> => {
        args.push(arg_trail);
        (l, WidgetCall::new(name, args, Vec::new()), r).into()
    },
};
```

The problem with this is obvious, first, the parser does not work if there are any children. Second, the missing value detection only works on the last argument. 

I've tried working around this for quite a long time now, but I only get more issues with ambiguous grammars and multiple productions. What would be the best way to solve something like this?

Runnable code (including some tests): https://github.com/Elias-Graf/rotten-potato/commit/45056f772d2bfa05daa7694d0e89e936c6a6d7eb

I've also considered making the error reporting part of a post-processing step. Although I'd really prefer not to, since it's syntax, and it would make interpreting the AST generally more complicated. But even then, as long as I make the atom in the argument optional, I will run into the same problem as far as I'm concerned.

**Comment by yannham** [comment]
I think I have a hard time understanding, even as humans, how one can distinguish between both cases. Let's say you remove the `1` from your example, you get:

```
(metric 
    :foo "bar"
    :label "123"
    :value
    "hello, world!"
    (bar))
```

But this can be interpreted as `:value` being `"hello, world!"` (just reformatting to make it clearer):

```
(metric 
    :foo "bar"
    :label "123"
    :value "hello, world!"
    (bar))
```

Do you have one or more examples of the kind of errors that you'd like to catch (missing value for an argument) ? With the corresponding source?

**Reply by Elias-Graf** [reply]
I'm not sure what you are asking. What I would want in this case? I think it would be perfectly sane to assume that the string was intended as the argument, no? I'm not even sure how you would reasonably catch this. Even if you somehow made the grammar whitespace sensitive, I don't think the specification forbids:

```yuck
(metric 
    :foo "bar"
    :label "123"
    :value
       "really long string on seperate line"
    (bar))
```

I want to provide the best possible messages, but within reason, for example:

```
(metric 
    :foo "bar"
    : "123" // missing parameter name
    :value "hello, world!"
    (bar))
```

```
(metric 
    :foo "bar"
    :label  // missing value
    :value "hello, world!"
    (bar))
```

```
(metric 
    :foo "bar"
    label "123" // incorrect parameter definition - missing colon
    :value "hello, world!"
    (bar))
```

I'm totally fine with doing "the most reasonable thing" in case of ambiguity. I maybe wasn't clear about that in the post. I think the snippets above are the main use cases I would like to cover. Probably good to also handle a combination of them, so if you start typing:

```
(metric 
    :)
```

Here the most sensible message would probably be that the name is missing, and then continue on to: the value is missing.

**Reply by yannham** [reply]
Thanks, it's exactly what I was asking: typical examples of what you want to handle/detect. I must admit that right now, I haven't given a lot of thought to your specific problem, but my general understaning of parser generators such as LALRPOP is that they aren't very good for fine-grained error messages like this. They are good at parsing valid syntax, but trying to encode many different ways of parsing incorrect input to account for various errors is usually painful, because of the non-ambiguity constraints (the parser needs to know with limited information if it should shift or reduce).

I'm not saying that there isn't a clever way to handle it in your case. But, if that were me, I think I would rather parse arguments in a proto structure (not the AST - another one that you only use in the parser) that is much more permissive, and then perform validation as part of the action code to turn this into a well-formed AST (could just be an implementation of `std::convert::TryFrom`), and raise a nice error if something is wrong. You can embed all the position information you need in the proto-AST when parsing, and raise custom LALRPOP errors to bubble the issue up.

For example, in Nickel, we have macros like that that automatically apply a fallible conversion function to their arguments and raise a custom LARLPOP error if something goes wrong: https://github.com/tweag/nickel/blob/026e43bb381069a0fe19bd7e4017121a014cbc8c/core/src/parser/grammar.lalrpop#L103. The error can be handled like any other parsing error by the caller.

Just my 2 cents.

**Reply by Elias-Graf** [reply]
Yes, I was afraid of something like this. Given that I currently would prefer not to switch the parsing framework, how would that "Proto-Grammar" look like? I still have the issue, that there are potentially two consecutive, optional atoms. Or am I missing something, because I'm tunnel visioning on my specific use-case?

**Reply by yannham** [reply]
The proto-AST doesn't have to be very complicated. There are probably many other possible designs, but here is how I would go about it as a first try:

1. Have a separate lexeme/token for argument name like `:foo`. Otherwise, unless you use a very specific LALRPOP setup, writing something like `":" <Symbol>` will both match `: foo` and `:foo` because LALRPOP ignores whitespaces between tokens, so you won't be able to give nice error messages for missing parameter name. Let's call it this token (uncreatively) `ColonSymbol`. Let's also say that you require at least one character after the starting `:`, to differentiate it from the single colon (it's just a convention - I think it could work as well to merge the colon in the same token as an empty parameter name, if you don't use the colon elsewhere).
2. Create an enum that can hold all proto-widget call items, that is
    ```rust
    enum WidgetArgItem {
      Name(String),
      ValueOrChild(YourAstType),
    }
    ```
3. Create a corresponding rule for such items, that should now be unambiguous for the parser. You can already handle a missing parameter name without post-processing:
    ```lalrpop
    pub WidgetArgItem: WidgetArgItem = {
      ColonSymbol => WidgetArgItem::Name(<>),
      <Atom | WidgetCall> => WidgetArgItem::ValueOrChild(<>),     
      ":" =>? Err(ParseError::User {
          error: MyError::MissingParamName,
      }),
    };
    ```
4. Parse arguments simply as a list of items, and then post-process:
    ```lalrpop
    pub WidgetCall: Spanned<WidgetCall> = {
        "(" <name:Symbol> <args: WidgetArgItem*> ")" =>? {
            WidgetCall::from_items(name, args).map_err(|error| ParseError::User {error})
        },
    };
    ```

In this case, `from_items` would be fallible function returning `Result<WidgetCall, E>` that can do the second stage parsing, which I think here would be to make sure that the vector of items starts with (a potentially empty) prefix of alternating `Name` and `ValueOrChild`s, and then a sequence of `ValueOrChild`, and put that in your `WidgetCall` structure. You can look at the [fallible action section](https://lalrpop.github.io/lalrpop/tutorial/007_fallible_actions.html) of the book to see how to make custom errors that can be embedded in LALRPOP parse errors as I do here.

There are probably multiple ways of going about this, and I might have overlooked the specifics of your case that would entail a bit of tweaking. But hopefully this conveys the general idea.

**Reply by yannham** [reply]
(So, to answer your initial question, while "proto-AST" can sound like a whole different AST, here it would just be a short-lived `Vec<Param name or atom or widget call>` with no constraints on their particular order that is converted away to the real AST directly in the `WidgetCall` rule and thus would be invisible from the outside otherwise)

**Reply by Elias-Graf** [reply]
So essentially just throwing everything a widget can contain into a single parser, and check order and validity, not using grammar? I think I'm going to try this, but it also feels like it is removing a lot of the value, of having a "grammar-framework" in the first place.

# The Section I wrote before, _I think_, I realized what you meant:

I don't think I can follow. The enum: `WidgetArgItem`, might be a value or a child? If we take a look back at the original grammar I posted:

```
pub WidgetCall: Spanned<WidgetCall> = {
    <l:@L> "(" <name:Symbol> <args:WidgetCallArg*> <children:WidgetCallChild*> ")" <r:@R> => {
        (l, WidgetCall::new(name, args, children), r).into()
    },
};
```

The arguments and children are two different consecutive structures, not something interlaced. What I tried to ask in my previous question, was essentially, how to write a parser that has:

```
<arg:WidgetCallArg> <child:WidgetCallChild>
```

When `arg` might be an `Atom?`, and `child` might be an `Atom?`, we would have `Atom? Atom?`, thus the parser wouldn't be able to decide if it should reduce to an `arg`, or shift to a `child`.

**Reply by yannham** [reply]
> So essentially just throwing everything a widget can contain into a single parser, and check order and validity, not using grammar? I think I'm going to try this, but it also feels like it is removing a lot of the value, of having a "grammar-framework" in the first place.

Well, the grammar is still useful for parsing e.g. nested widget calls, parentheses, basic values. I would say there's just _part_ of the grammar that is indeed offloaded to post-processing step.

In general, LALR parsers (and more generally generated parsers) are one trade-off in the design space. They are good at getting a grammar up and running quickly, at making the grammar "easy" to read (to map to and from an EBNF formal description, for example), and at making this grammar simple/easy to extend (as well as handling some other aspects such as precedence which isn't trivial in e.g. a basic handwritten recursive descent parser). They are reasonably fast. You can also write your parser in a left recursive style, which is much more natural.

However, LALR parsers aren't magical. It' more difficult to give very nice and specialized error messages (well, depends compared to _what_ --but I think this is fair to say when compared to recursive descent parsers). They aren't necessarily always the fastest solution. I believe a number of mainstream language implementations use recursive descent parsing once the grammar is stable enough precisely to give nice error messages.

All in all, I think if widget stuff is just one part of your grammar, with many other rules that don't need such post-processing, this is still worth it. Or if you plan to experiment a bit with the syntax, extending or changing things along the way. On the other hand, if your grammar is small (mostly widget calls), fixed, and error messages are really important for you, then maybe LALRPOP is a bit superfluous, and you could do as much by just lexing and post-processing, or by using another parsing framework which is particularly good at error messages.

P.S:  another possible technique would be to separate the fast and happy path, that is the source is valid input, and the error path. For example it's not hard to include error recovery in a LALRPOP parser with the special rule `!`, so that instead of erroring out, you can embed a special `Error` node in your AST. Doing so, you could keep your parser simple, without caring about error messages but focusing just on describing the grammar (no rule to match invalid widget arguments). Then, if you find an error node (that you would store with position information) in place of widget arguments for example, you could re-run a special, different rule, or even just an entirely different lexer that don't need to use the LALRPOP parser at all and can try to say something smart about the error. That would keep the main LALRPOP grammar clean, at least, but would probably lead to more code.
